### PR TITLE
Refactor block click and break handling.

### PR DESCRIPTION
**Pull request by bloodmc**
Currently, block clicks and breaks are scheduled to run the following tick
by sending a custom packet to server which performs the respective action.
This causes all sorts of issues with mods such as Sponge that expect the
break and click action to occur the same tick as the player action. In
addition, both the vanilla packet and custom packet are sent each time a player
performs a break or click on a storage block.

This commit aims at refactoring this code entirely to follow vanilla standards
and achieve higher compatibility with other mods especially ones that
provide protection.

Here are the bulk of the changes
1. Instead of sending the 'interactShift' change everytime a player clicks
   a block on client, this is now handled via the 'BoolConfigUpdateMessage'
   packet. When a player joins the server for the first time, the
   'interactShift' value will be sent to the server and stored in memory
   until the player disconnects. If the player opens up the config gui and
   changes it while in game, the same packet will be sent to server to update
   value stored or add it as new.
2. Instead of overriding onBlockClicked just to pass hitVec, rayTracing
   will also occur on server-side by using Forge's built-in hooks. This
   allows the standard onBlockClicked method to be used and avoids headaches
   with future updates based on how it is coded currently.
3. Scheduled tasks have been removed entirely as there is no longer a
   reason for them. Once a player clicks or breaks a block, the respective
   action will occur immediately. Due to this, the BlockClickMessage and
   BlockDestroyMessage classes have been removed.

### CODE REVIEW

File: src/com/jaquadro/minecraft/storagedrawers/block/BlockDrawers.java [hunk 1]
```diff
@@ -415,94 +413,66 @@ protected boolean hitLeft (int side, float hitX, float hitZ) {
...
+        RayTraceResult rayResult = net.minecraftforge.common.ForgeHooks.rayTraceEyes(playerIn, ((EntityPlayerMP) playerIn).interactionManager.getBlockReachDistance() + 1);
+        EnumFacing side = rayResult.sideHit;
+        // adjust hitVec for drawers
+        float hitX = (float)(rayResult.hitVec.xCoord - pos.getX());
+        float hitY = (float)(rayResult.hitVec.yCoord - pos.getY());
+        float hitZ = (float)(rayResult.hitVec.zCoord - pos.getZ());
 
-        if (!player.capabilities.isCreativeMode) {
-            PlayerInteractEvent.LeftClickBlock event = new PlayerInteractEvent.LeftClickBlock(player, pos, side, new Vec3d(hitX, hitY, hitZ));
+        if (!playerIn.capabilities.isCreativeMode) {
```
Comment (Zidane) [inline src/com/jaquadro/minecraft/storagedrawers/block/BlockDrawers.java]: Should Creative still fire LeftClickBlock?

Comment (ghost) [reply src/com/jaquadro/minecraft/storagedrawers/block/BlockDrawers.java]: tested this patch with spongeforge, but issue with duplication in the wilderness still persist

**Comment by bloodmc** [comment]
@hunter688 Are you testing this in creative? From my tests, the item is NOT duplicated, it only shows up in the container as a visual since the client doesn't know the action was cancelled. I guess I could send a packet to client to update the visual but it's not really a big deal.

File: src/com/jaquadro/minecraft/storagedrawers/block/BlockDrawers.java (hunk 1 above)
Comment (bloodmc) [reply src/com/jaquadro/minecraft/storagedrawers/block/BlockDrawers.java]: @Zidane I'll remove it but there is a case where onBlockClicked is called if a player attempts to break the front side of a storage block. However, you are correct that the event should fire normally for the actual break so it shouldn't be needed. I'll test it tonight.

**Comment by ghost** [comment]
tested in the survival as normal player in the wilderness and still can dupe

**Comment by bloodmc** [comment]
@hunter688 dupe has been fixed in latest SF build.

**Comment by jaquadro** [comment]
Oh cool.  I think I need to start instructing people to @ mention me in PRs because I keep overlooking them (work repos drown my notifications).

I'll try and look closer in the next couple days.  I won't be sad to see the client click message go along with its various problems.

**Comment by bloodmc** [comment]
@jaquadro Any luck with this PR?

**Comment by jaquadro** [comment]
I've reviewed the code.  It looks good.  I've just had interruptions every time I try to sit down and work on mods.  Hopefully this weekend.

**Pull request merged**